### ISSUE TITLE
Run nff-go on Azure

### ISSUE THREAD

**Issue by mkfsn** [intel-go/nff-go#467]
Hi, I found a [post](https://docs.microsoft.com/en-us/azure//virtual-network/setup-dpdk) to run DPDK on Azure Linux virtual machine.

I'm using Ubuntu 16.04 with kernel 4.15.0-1015-azure, DPDK 18.02, and nff-go 0.4.1.

In the post, I also found that I have to enable the mellanox PMDs:

```
sed -ri 's,(MLX._PMD=)n,\1y,' build/.config
```

and then I can run the testpmd successfully:

```
./build/app/testpmd -w 0002:00:02.0 --vdev="net_vdev_netvsc0,iface=eth1" -- -i --port-topology=chained
```

However, I have no idea how to run nff-go on Azure, do you have any suggestion how to configure nff-go?

Right now I've tried setting the DPDKArgs: 
- []string{"--vdev", "net_vdev_netvsc0,iface=eth1"}
- []string{"--vdev", "net_vdev_netvsc0,iface=eth1", "-w", "0002:00:02.0"}

but still doesn't work

---

Also, here is my environment:

- devbind
```
Network devices using DPDK-compatible driver
============================================
<none>

Network devices using kernel driver
===================================
0002:00:02.0 'MT27500/MT27520 Family [ConnectX-3/ConnectX-3 Pro Virtual Function] 1004' if=enP2p0s2 drv=mlx4_core unused=

Other Network devices
=====================
<none>

Crypto devices using DPDK-compatible driver
===========================================
<none>

Crypto devices using kernel driver
==================================
<none>

Other Crypto devices
====================
<none>

Eventdev devices using DPDK-compatible driver
=============================================
<none>

Eventdev devices using kernel driver
====================================
<none>

Other Eventdev devices
======================
<none>

Mempool devices using DPDK-compatible driver
============================================
<none>

Mempool devices using kernel driver
===================================
<none>

Other Mempool devices
=====================
<none>
```

- ifconfig

```
enP2p0s2  Link encap:Ethernet  HWaddr 00:0d:3a:19:f1:30
          UP BROADCAST RUNNING SLAVE MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:9 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:1510 (1.5 KB)

eth0      Link encap:Ethernet  HWaddr 00:0d:3a:13:12:24
          inet addr:10.240.0.101  Bcast:10.240.255.255  Mask:255.255.0.0
          inet6 addr: fe80::20d:3aff:fe13:1224/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:430248 errors:0 dropped:0 overruns:0 frame:0
          TX packets:275592 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:468150790 (468.1 MB)  TX bytes:42800771 (42.8 MB)

eth1      Link encap:Ethernet  HWaddr 00:0d:3a:19:f1:30
          inet addr:10.240.254.103  Bcast:10.240.255.255  Mask:255.255.0.0
          inet6 addr: fe80::20d:3aff:fe19:f130/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:2 errors:0 dropped:0 overruns:0 frame:0
          TX packets:9 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:762 (762.0 B)  TX bytes:1510 (1.5 KB)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:178 errors:0 dropped:0 overruns:0 frame:0
          TX packets:178 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:12560 (12.5 KB)  TX bytes:12560 (12.5 KB)
```

- ethtool -i enP2p0s2

```
driver: mlx4_en
version: 4.0-0
firmware-version: 2.41.7004
expansion-rom-version:
bus-info: 0002:00:02.0
supports-statistics: yes
supports-test: yes
supports-eeprom-access: no
supports-register-dump: no
supports-priv-flags: yes
```

- lspci

```
0000:00:00.0 Host bridge: Intel Corporation 440BX/ZX/DX - 82443BX/ZX/DX Host bridge (AGP disabled) (rev 03)
0000:00:07.0 ISA bridge: Intel Corporation 82371AB/EB/MB PIIX4 ISA (rev 01)
0000:00:07.1 IDE interface: Intel Corporation 82371AB/EB/MB PIIX4 IDE (rev 01)
0000:00:07.3 Bridge: Intel Corporation 82371AB/EB/MB PIIX4 ACPI (rev 02)
0000:00:08.0 VGA compatible controller: Microsoft Corporation Hyper-V virtual VGA
0002:00:02.0 Ethernet controller: Mellanox Technologies MT27500/MT27520 Family [ConnectX-3/ConnectX-3 Pro Virtual Function]
```

Any advice and suggestions will be greatly appreciated.

**Comment by gshimansky** [comment]
If you can successfully build NFF-Go and bind a port to DPDK driver, just use NFF-Go with port index 0 (the first one), and it should work.

**Comment by mkfsn** [comment]
@gshimansky thank you for the quick reply

I actually get this error while setting receive port:

```
Failed to set receiver: Requested receive port exceeds number of ports which can be used by DPDK (bind to DPDK). (3)
```

because I'm getting 0 from 

https://github.com/intel-go/nff-go/blob/9a5e68c5e09db53d970108ccf081251dfa7e9b7d/low/low.go#L511


If I read the [post](https://docs.microsoft.com/en-us/azure//virtual-network/setup-dpdk) correcly, I don't need to bind any DPDK driver to the _VF_ network interface. 

Or maybe I have some misunderstandings - at least I'm seeing no changes in devbind status, before and after I run the testpmd:

- before
```
Network devices using kernel driver
===================================
0002:00:02.0 'MT27500/MT27520 Family [ConnectX-3/ConnectX-3 Pro Virtual Function] 1004' if=enP2p0s2 drv=mlx4_core unused=
```

- after

```
Network devices using kernel driver
===================================
0002:00:02.0 'MT27500/MT27520 Family [ConnectX-3/ConnectX-3 Pro Virtual Function] 1004' if=enP2p0s2 drv=mlx4_core unused=
```

Do you have any advice or suggestion?

**Comment by gshimansky** [comment]
You need to bind ports to DPDK driver because NFF-Go works only with ports bound to DPDK driver. `GetPortsNumber` calls [rte_eth_dev_count](http://doc.dpdk.org/api/rte__ethdev_8h.html#a49122554b164a59667d46d144eda2ce8) inside, and if this function returns 0, it means that there are no ethernet devices found that NFF-Go can use.

**Comment by gshimansky** [comment]
Unfortunately, I am not familiar with the way `testpmd` is used in the article that you are referring to. Usually I run `testpmd` with `--portmask=0xXX` and it DPDK bound ports by indexes starting with zero. Current implementation of NFF-Go requires ports to be bound to DPDK driver and they are also referred to by indexes starting with zero.
What output are you getting for the command `show port info all` which should do the configuration sanity check?

**Comment by mkfsn** [comment]
@gshimansky I see. Here is the output of `show port info all`

```
testpmd> show port info all

********************* Infos for port 0  *********************
MAC address: 00:0D:3A:19:F1:30
Driver name: net_mlx4
Connect to socket: 0
memory allocation on the socket: 0
Link status: up
Link speed: 40000 Mbps
Link duplex: full-duplex
MTU: 1500
Promiscuous mode: enabled
Allmulticast mode: disabled
Maximum number of MAC addresses: 128
Maximum number of MAC addresses of hash filtering: 0
VLAN offload:
  strip off
  filter off
  qinq(extend) off
Hash key size in bytes: 40
No flow type is supported.
Minimum size of RX buffer: 32
Maximum configurable length of RX packet: 65536
Current number of RX queues: 1
Max possible RX queues: 12288
Max possible number of RXDs per queue: 65535
Min possible number of RXDs per queue: 0
RXDs number alignment: 1
Current number of TX queues: 1
Max possible TX queues: 12288
Max possible number of TXDs per queue: 65535
Min possible number of TXDs per queue: 0
TXDs number alignment: 1
```

**Comment by mkfsn** [comment]
Unfortunately, I couldn't load igb_uio driver using kernel 4.15.0-1015-azure, looks like I need to recompile the kernel.

However, I can load igb_uio when using kernel 4.14.0-1003-azure-edge:

```
Network devices using DPDK-compatible driver
============================================
0002:00:02.0 'MT27500/MT27520 Family [ConnectX-3/ConnectX-3 Pro Virtual Function] 1004' drv=igb_uio unused=mlx4_core

Network devices using kernel driver
===================================
<none>

Other Network devices
=====================
<none>

Crypto devices using DPDK-compatible driver
===========================================
<none>

Crypto devices using kernel driver
==================================
<none>

Other Crypto devices
====================
<none>

Eventdev devices using DPDK-compatible driver
=============================================
<none>

Eventdev devices using kernel driver
====================================
<none>

Other Eventdev devices
======================
<none>

Mempool devices using DPDK-compatible driver
============================================
<none>

Mempool devices using kernel driver
===================================
<none>

Other Mempool devices
=====================
<none>
``` 

but I still get 0 from rte_eth_dev_count().

**Comment by gshimansky** [comment]
Looks like you successfully bound your port to DPDK, so it should now work with DPDK programs.
Do you see any error messages in `dmesg`?
What happens if you run `sudo testpmd` with no other arguments?

**Comment by mkfsn** [comment]
I didn't see any error messages in `dmesg` but in `journalctl`:

```
Aug 28 03:06:57 router router[7924]: ------------***-------- Initializing DPDK --------***------------
Aug 28 03:06:57 router router[7924]: EAL: Detected 2 lcore(s)
Aug 28 03:06:57 router router[7924]: EAL: No free hugepages reported in hugepages-1048576kB
Aug 28 03:06:57 router router[7924]: EAL: Multi-process socket /var/run/.rte_unix
Aug 28 03:06:57 router router[7924]: EAL: Probing VFIO support...
Aug 28 03:06:57 router router[7929]: EAL: Multi-process socket /var/run/.rte_unix
Aug 28 03:06:57 router router[7929]: EAL: Probing VFIO support...
Aug 28 03:06:58 router router[7924]: EAL: WARNING: cpu flags constant_tsc=yes nonstop_tsc=no -> using unreliable clock cycles !
Aug 28 03:06:58 router router[7929]: EAL: WARNING: cpu flags constant_tsc=yes nonstop_tsc=no -> using unreliable clock cycles !
Aug 28 03:06:58 router router[7929]: PMD: net_failsafe: Initializing Fail-safe PMD for net_failsafe_net_vdev_netvsc_id0
Aug 28 03:06:58 router router[7924]: PMD: net_failsafe: Initializing Fail-safe PMD for net_failsafe_net_vdev_netvsc_id0
Aug 28 03:06:58 router router[7924]: PMD: net_failsafe: Creating fail-safe device on NUMA socket 0
Aug 28 03:06:58 router router[7924]: ERROR: failed to parse device "net_tap_net_vdev_netvsc_id0"
Aug 28 03:06:58 router router[7924]: EAL: Driver cannot attach the device (net_failsafe_net_vdev_netvsc_id0)
Aug 28 03:06:58 router router[7924]: PMD: net_failsafe: Initializing Fail-safe PMD for net_failsafe_net_vdev_netvsc_id0
Aug 28 03:06:58 router router[7924]: PMD: net_failsafe: Creating fail-safe device on NUMA socket 0
Aug 28 03:06:58 router router[7924]: PMD: net_failsafe: This PMD requires sub-devices, none provided
Aug 28 03:06:58 router router[7924]: vdev_probe(): failed to initialize  device
Aug 28 03:06:58 router router[7924]: EAL: Bus (vdev) probe failed.
Aug 28 03:06:58 router router[7924]: ------------***------ Initializing scheduler -----***------------
Aug 28 03:06:58 router router[7924]: DEBUG: Scheduler can use cores: [0 1]
Aug 28 03:06:58 router router[7929]: PMD: net_failsafe: Creating fail-safe device on NUMA socket 0
Aug 28 03:06:58 router router[7929]: EAL: Driver cannot attach the device (net_failsafe_net_vdev_netvsc_id0)
Aug 28 03:06:58 router router[7929]: PMD: net_failsafe: Initializing Fail-safe PMD for net_failsafe_net_vdev_netvsc_id0
Aug 28 03:06:58 router router[7929]: PMD: net_failsafe: Creating fail-safe device on NUMA socket 0
Aug 28 03:06:58 router router[7929]: PMD: net_failsafe: This PMD requires sub-devices, none provided
Aug 28 03:06:58 router router[7929]: vdev_probe(): failed to initialize  device
Aug 28 03:06:58 router router[7929]: EAL: Bus (vdev) probe failed.
```

Looks like it's using `net_failsafe` PMD with a werid device name `net_failsafe_net_vdev_netvsc_id0`.


Also, if I run testpmd without any arguments:

```
EAL: Detected 2 lcore(s)
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Multi-process socket /var/run/.rte_unix
EAL: Probing VFIO support...
EAL: WARNING: cpu flags constant_tsc=yes nonstop_tsc=no -> using unreliable clock cycles !
PMD: net_failsafe: Initializing Fail-safe PMD for net_failsafe_net_vdev_netvsc_id0
PMD: net_failsafe: Creating fail-safe device on NUMA socket 0
PMD: Initializing pmd_tap for net_tap_net_vdev_netvsc_id0 as dtap0
PMD: net_failsafe: MAC address is 00:0d:3a:19:f1:30
Warning: NUMA should be configured manually by using --port-numa-config and --ring-numa-config parameters along with --numa.
testpmd: create a new mbuf pool <mbuf_pool_socket_0>: n=155456, size=2176, socket=0
testpmd: preferred mempool ops selected: ring_mp_mc

Warning! port-topology=paired and odd forward ports number, the last port will pair with itself.

Configuring Port 0 (socket 0)
PMD: net_tap_net_vdev_netvsc_id0: 0xf85e40: TX configured queues number: 1
PMD: net_tap_net_vdev_netvsc_id0: 0xf85e40: RX configured queues number: 1
Port 0: 00:0D:3A:19:F1:30
Checking link statuses...
Done
No commandline core given, start packet forwarding
io packet forwarding - ports=1 - cores=1 - streams=1 - NUMA support enabled, MP over anonymous pages disabled
Logical Core 1 (socket 0) forwards packets on 1 streams:
  RX P=0/Q=0 (socket 0) -> TX P=0/Q=0 (socket 0) peer=02:00:00:00:00:00

  io packet forwarding packets/burst=32
  nb forwarding cores=1 - nb forwarding ports=1
  port 0:
  CRC stripping enabled
  RX queues=1 - RX desc=1024 - RX free threshold=0
  RX threshold registers: pthresh=0 hthresh=0  wthresh=0
  TX queues=1 - TX desc=1024 - TX free threshold=0
  TX threshold registers: pthresh=0 hthresh=0  wthresh=0
  TX RS bit threshold=0 - TXQ offloads=0x0
Press enter to exit
```

It seems to me the biggest difference is this line:

```diff
- ERROR: failed to parse device "net_tap_net_vdev_netvsc_id0"
+ PMD: Initializing pmd_tap for net_tap_net_vdev_netvsc_id0 as dtap0
```

**Comment by aregm** [comment]
https://doc.dpdk.org/guides/linux_gsg/sys_reqs.html - for reference.

**Comment by mkfsn** [comment]
@aregm I did already
```
# cat /sys/devices/system/node/node0/hugepages/hugepages-2048kB/nr_hugepages
1024
```

**Comment by gshimansky** [comment]
What version of DPDK is this? I think there is some problem in DPDK interacting with Azure environment that it fails to initialize a driver when binding a port. I can only recommend to as on DPDK mail list about it.

**Comment by mkfsn** [comment]
@gshimansky I'm using DPDK 18.02, as Azure required:

> Version 18.02 or higher is required for Azure

Also, I'm using nff-go 0.3.

**Comment by marcusschiesser** [comment]
Same issue still exists with nff-go 0.7

**Comment by guesslin** [comment]
Bug about DPDK on Azure using `intel-go/nff-go` fails using `hv_netvsc` driver
https://bugs.dpdk.org/show_bug.cgi?id=175

**Comment by shemminger** [comment]
Let's separate this into two issues: DPDK support with vdev_netvsc, and DPDK support with hv_netvsc.

The biggest is that NFF-go is not built with support for Mellanox in the DPDK. This is because the Makefile depends on DPDK config file. There is no DPDK defconfig with the necessary Mellanox driver.
If that was fixed vdev_netvsc would work.

The hv_netvsc has a different rebinding (ie not PCI) and the NFF-go infrastructure is not aware of VMBus.  Like VPP changes are probably necessary to do this.

**Comment by gshimansky** [comment]
MLX4 support in NFF-Go is something that we have in plans, but since it also has dependencies on `libvers` and `mlx4` Linux libraries, we didn't add it to the default build. I am thinking of making an optional environment variable like `ENABLE_MLX4` which would enable all dependencies in NFF-Go build system and link MLX4 PMD driver into NFF-Go binaries. What do you think about this approach?

**Comment by shemminger** [comment]
On Fri, 04 Jan 2019 10:18:46 -0800
Gregory Shimansky <notifications@github.com> wrote:

> MLX4 support in NFF-Go is something that we have in plans, but since it also has dependencies on `libvers` and `mlx4` Linux libraries, we didn't add it to the default build. I am thinking of making an optional environment variable like `ENABLE_MLX4` which would enable all dependencies in NFF-Go build system and link MLX4 PMD driver into NFF-Go binaries. What do you think about this approach?
> 

I was thinking of getting DPDK to accept a config file which was Azure specific.
For Azure you want mlx5 and rdma-core.

Alternatively, you could do what does VPP and use DPDK_MLX5_PMD=y on the make command line to enable it.
They generate their own DPDK config file.

**Comment by shemminger** [comment]
Another option is to use distribution provided library for DPDK.

DPDK is transitioning to meson build which auto detects available
dependency. Maybe use that in the future?

**Comment by gshimansky** [comment]
There is a Go specific problem because we have to specify `CGO_LDFLAGS` in a `.go` source file. Go supports "conditional compilation" with `-tag` parameter, but I am not sure how `CGO_LDGLAGS` from multiple files are going to be concatenated.

**Comment by gshimansky** [comment]
I found a guide on binding netvsc interfaces to DPDK https://doc.dpdk.org/guides/nics/netvsc.html . I created two additional network interfaces on my Azure VM, e.g.
```
ethtool -i eth2
driver: hv_netvsc
version:
firmware-version: N/A
expansion-rom-version:
bus-info:
supports-statistics: yes
supports-test: no
supports-eeprom-access: no
supports-register-dump: no
supports-priv-flags: no
```
and tried to follow the steps to bind it to `uio_hv_generic` driver. I have to do it manually because in Azure both Ubuntu 16 and Ubuntu 18 have kernels 4.15.x. But something still doesn't work for me because when I run testpmd it fails to find these interfaces:
```
./testpmd -l 0-4 -- -i
EAL: Detected 8 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
EAL: WARNING: cpu flags constant_tsc=yes nonstop_tsc=no -> using unreliable clock cycles !
get_sysfs_string(): cannot open sysfs value /sys/bus/vmbus/devices/000d3a31-b538-000d-3a31-b538000d3a31/uio/uio0/maps/map3/name:No such file or directory
rte_vmbus_map_device(): could not read /sys/bus/vmbus/devices/000d3a31-b538-000d-3a31-b538000d3a31/uio/uio0/maps/map3/name
rte_vmbus_probe(): Requested device 000d3a31-b538-000d-3a31-b538000d3a31 cannot be used
get_sysfs_string(): cannot open sysfs value /sys/bus/vmbus/devices/000d3a38-09da-000d-3a38-09da000d3a38/uio/uio1/maps/map3/name:No such file or directory
rte_vmbus_map_device(): could not read /sys/bus/vmbus/devices/000d3a38-09da-000d-3a38-09da000d3a38/uio/uio1/maps/map3/name
rte_vmbus_probe(): Requested device 000d3a38-09da-000d-3a38-09da000d3a38 cannot be used
testpmd: No probed ethernet devices
Interactive-mode selected
testpmd: create a new mbuf pool <mbuf_pool_socket_0>: n=179456, size=2176, socket=0
testpmd: preferred mempool ops selected: ring_mp_mc
Done
testpmd>
Bye...
```
I am not sure what `testpmd` is looking for in this directory, I have only `map0`, `map1` and `map2`:
```
ls -l /sys/bus/vmbus/devices/000d3a38-09da-000d-3a38-09da000d3a38/uio/uio1/maps
total 0
drwxr-xr-x 2 root root 0 Jan  9 21:59 map0
drwxr-xr-x 2 root root 0 Jan  9 21:59 map1
drwxr-xr-x 2 root root 0 Jan  9 21:59 map2
```

**Comment by aregm** [comment]
I think you need kernel 4.18.

**Comment by gshimansky** [comment]
I did find 4.18-azure kernel for Ubuntu 18.4
```
Linux ubuntu18-2 4.18.0-1007-azure #7~18.04.1-Ubuntu SMP Fri Dec 7 01:51:34 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
and indeed I managed to bind netvsc interface to uio driver with the same `echo` commands from netvsc guide. I even managed to see interface information with `testpmd`!
```
root@ubuntu18-2:/home/gashiman/nff-go/dpdk/dpdk/x86_64-native-linuxapp-gcc/app# ./testpmd -l 0-4 -- -i
EAL: Detected 8 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
EAL: WARNING: cpu flags constant_tsc=yes nonstop_tsc=no -> using unreliable clock cycles !
vmbus_probe_one_driver():   Invalid NUMA socket, default to 0
Interactive-mode selected
testpmd: create a new mbuf pool <mbuf_pool_socket_0>: n=179456, size=2176, socket=0
testpmd: preferred mempool ops selected: ring_mp_mc

Warning! port-topology=paired and odd forward ports number, the last port will pair with itself.

Configuring Port 0 (socket 0)
Port 0: 00:0D:3A:31:B5:38
Checking link statuses...
Done
testpmd> show port info all

********************* Infos for port 0  *********************
MAC address: 00:0D:3A:31:B5:38
Device name: 000d3a31-b538-000d-3a31-b538000d3a31
Driver name: net_netvsc
Connect to socket: 0
memory allocation on the socket: 0
Link status: down
Link speed: 0 Mbps
Link duplex: half-duplex
MTU: 1500
Promiscuous mode: enabled
Allmulticast mode: disabled
Maximum number of MAC addresses: 1
Maximum number of MAC addresses of hash filtering: 0
VLAN offload:
  strip off
  filter off
  qinq(extend) off
Hash key size in bytes: 40
Supported RSS offload flow types:
  ipv4
  ipv4-tcp
  ipv4-udp
  ipv6
  ipv6-tcp
  ipv6-udp
  user defined 16
  user defined 17
Minimum size of RX buffer: 1024
Maximum configurable length of RX packet: 2048
Current number of RX queues: 1
Max possible RX queues: 64
Max possible number of RXDs per queue: 65535
Min possible number of RXDs per queue: 0
RXDs number alignment: 1
Current number of TX queues: 1
Max possible TX queues: 64
Max possible number of TXDs per queue: 65535
Min possible number of TXDs per queue: 0
TXDs number alignment: 1
testpmd>
Shutting down port 0...
Stopping ports...
Done
Closing ports...
Done

Bye...
```
Unfortunately it worked only once for me. After I started `testpmd` a second time it hanged and the whole system had to be restarted in Azure console.

**Comment by shemminger** [comment]
On Wed, 09 Jan 2019 22:20:27 +0000 (UTC)
Gregory Shimansky <notifications@github.com> wrote:

> Unfortunately it worked only once for me. After I started `testpmd` a second time it hanged and the whole system had to be restarted in Azure console.

The uio_hv_generic kernel driver doesn't handle restart correctly until the most recent
kernels.  A fix for restart went into 4.20 but that also introduced another issue
with multi-queue; the fix for that issue was submitted but lost in the transition
to different kernel process for Hyper-V.

**Comment by aregm** [comment]
@shemminger Stephen, do you have advice here?

**Comment by gshimansky** [comment]
@shemminger thank you for your insight. Do you by chance know if it is possible to use vanilla kernels from kernel.org on Azure VMs or some specific patches are required? Or is there some other way to run 4.20.x in Azure?

For Mellanox driver we decided to include it into NFF-Go build unconditionally which will result in new dependencies (libibverbs-dev) but will allow running NFF-Go with any Mellanox cards with no modifications. I am going to do a corresponding PR later today.

**Comment by shemminger** [comment]
On Thu, 10 Jan 2019 07:47:57 -0800
Gregory Shimansky <notifications@github.com> wrote:

> @shemminger thank you for your insight. Do you by chance know if it is possible to use vanilla kernels from kernel.org on Azure VMs or some specific patches are required? Or is there some other way to run 4.20.x in Azure?
> 
> For Mellanox driver we decided to include it into NFF-Go build unconditionally which will result in new dependencies (libibverbs-dev) but will allow running NFF-Go with any Mellanox cards with no modifications. I am going to do a corresponding PR later today.
> 

You also need rdma-core.

I have started automating the setup of testbed on Azure, will add nff-go installation script as well.

Note: using native netvsc device is still experimental at this time. The recommended solution
is to use vdev_netvsc which setups up TAP/failsafe and VF.

**Comment by gshimansky** [comment]
Mellanox network drivers should be built now in DPDK which NFF-Go links against. I integrated PR #533 into devel branch.

**Comment by gshimansky** [comment]
Yes with updated DPDK build in `develop` branch I could successfully bind two Azure network ports to failsafe driver with the following command:
```
gashiman@ubuntu18-2:~/nff-go/dpdk/dpdk/x86_64-native-linuxapp-gcc-install/usr/local/bin$ sudo ./testpmd -w 0000:00:08.0 --vdev="net_vdev_netvsc0,iface=eth1" --vdev="net_vdev_netvsc1,iface=eth2" -- -i
EAL: Detected 8 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
EAL: WARNING: cpu flags constant_tsc=yes nonstop_tsc=no -> using unreliable clock cycles !
net_vdev_netvsc: probably using routed NetVSC interface "eth1" (index 3)
rte_pmd_tap_probe(): Initializing pmd_tap for net_tap_vsc0 as dtap0
net_vdev_netvsc: probably using routed NetVSC interface "eth2" (index 4)
rte_pmd_tap_probe(): Initializing pmd_tap for net_tap_vsc1 as dtap1
Interactive-mode selected
Warning: NUMA should be configured manually by using --port-numa-config and --ring-numa-config parameters along with --numa.
testpmd: create a new mbuf pool <mbuf_pool_socket_0>: n=203456, size=2176, socket=0
testpmd: preferred mempool ops selected: ring_mp_mc
Configuring Port 0 (socket 0)
Port 0: 00:0D:3A:31:B5:38
Configuring Port 2 (socket 0)
Port 2: 00:0D:3A:38:09:DA
Checking link statuses...
Done
testpmd> show port info all

********************* Infos for port 0  *********************
MAC address: 00:0D:3A:31:B5:38
Device name: net_failsafe_vsc0
Driver name: net_failsafe
Devargs: remote=eth1
Connect to socket: 0
memory allocation on the socket: 0
Link status: up
Link speed: 10000 Mbps
Link duplex: full-duplex
MTU: 1500
Promiscuous mode: enabled
Allmulticast mode: disabled
Maximum number of MAC addresses: 1
Maximum number of MAC addresses of hash filtering: 0
VLAN offload:
  strip off
  filter off
  qinq(extend) off
Hash key size in bytes: 40
Supported RSS offload flow types:
  ipv4
  ipv4-frag
  ipv4-tcp
  ipv4-udp
  ipv4-other
  ipv6
  ipv6-frag
  ipv6-tcp
  ipv6-udp
  ipv6-other
  user defined 15
  user defined 16
  user defined 17
Minimum size of RX buffer: 0
Maximum configurable length of RX packet: 1522
Current number of RX queues: 1
Max possible RX queues: 16
Max possible number of RXDs per queue: 65535
Min possible number of RXDs per queue: 0
RXDs number alignment: 1
Current number of TX queues: 1
Max possible TX queues: 16
Max possible number of TXDs per queue: 65535
Min possible number of TXDs per queue: 0
TXDs number alignment: 1

********************* Infos for port 2  *********************
MAC address: 00:0D:3A:38:09:DA
Device name: net_failsafe_vsc1
Driver name: net_failsafe
Devargs: remote=eth2
Connect to socket: 0
memory allocation on the socket: 0
Link status: up
Link speed: 10000 Mbps
Link duplex: full-duplex
MTU: 1500
Promiscuous mode: enabled
Allmulticast mode: disabled
Maximum number of MAC addresses: 1
Maximum number of MAC addresses of hash filtering: 0
VLAN offload:
  strip off
  filter off
  qinq(extend) off
Hash key size in bytes: 40
Supported RSS offload flow types:
  ipv4
  ipv4-frag
  ipv4-tcp
  ipv4-udp
  ipv4-other
  ipv6
  ipv6-frag
  ipv6-tcp
  ipv6-udp
  ipv6-other
  user defined 15
  user defined 16
  user defined 17
Minimum size of RX buffer: 0
Maximum configurable length of RX packet: 1522
Current number of RX queues: 1
Max possible RX queues: 16
Max possible number of RXDs per queue: 65535
Min possible number of RXDs per queue: 0
RXDs number alignment: 1
Current number of TX queues: 1
Max possible TX queues: 16
Max possible number of TXDs per queue: 65535
Min possible number of TXDs per queue: 0
TXDs number alignment: 1
testpmd>
```
Two things of note about it.

1. I have to specify `-w` option with a PCI ID of a card that **is not network card!** because otherwise testpmd binds my ssh interface and I lose command line. Above I specified PCI ID of a VGA adapter, so `testpmd` binds only ports specified on command line. If I don't spcify `-w`, `testmpd` binds a real PCI network card (since now DPDK is built with Mellanox driver included) and prevents any other packets from being transferred over my SSH connection.
2. For some reason second port got index `2` instead of `1`. Not sure why it happened.

I am inclined to close this issue because it became quite long already and I think most if not all of the problems were resolved. If you have any objections please commend or I will close it on the next week.

**Comment by shemminger** [comment]
On Fri, 11 Jan 2019 10:57:49 -0800
Gregory Shimansky <notifications@github.com> wrote:

> Yes with updated DPDK build in `develop` branch I could successfully bind two Azure network ports to failsafe driver with the following command:
> ```
> gashiman@ubuntu18-2:~/nff-go/dpdk/dpdk/x86_64-native-linuxapp-gcc-install/usr/local/bin$ sudo ./testpmd -w 0000:00:08.0 --vdev="net_vdev_netvsc0,iface=eth1" --vdev="net_vdev_netvsc1,iface=eth2" -- -i
> EAL: Detected 8 lcore(s)
> EAL: Detected 1 NUMA nodes
> EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
> EAL: No free hugepages reported in hugepages-1048576kB
> EAL: Probing VFIO support...
> EAL: WARNING: cpu flags constant_tsc=yes nonstop_tsc=no -> using unreliable clock cycles !
> net_vdev_netvsc: probably using routed NetVSC interface "eth1" (index 3)
> rte_pmd_tap_probe(): Initializing pmd_tap for net_tap_vsc0 as dtap0
> net_vdev_netvsc: probably using routed NetVSC interface "eth2" (index 4)
> rte_pmd_tap_probe(): Initializing pmd_tap for net_tap_vsc1 as dtap1
> Interactive-mode selected
> Warning: NUMA should be configured manually by using --port-numa-config and --ring-numa-config parameters along with --numa.
> testpmd: create a new mbuf pool <mbuf_pool_socket_0>: n=203456, size=2176, socket=0
> testpmd: preferred mempool ops selected: ring_mp_mc
> Configuring Port 0 (socket 0)
> Port 0: 00:0D:3A:31:B5:38
> Configuring Port 2 (socket 0)
> Port 2: 00:0D:3A:38:09:DA
> Checking link statuses...
> Done
> testpmd> show port info all  
> 
> ********************* Infos for port 0  *********************
> MAC address: 00:0D:3A:31:B5:38
> Device name: net_failsafe_vsc0
> Driver name: net_failsafe
> Devargs: remote=eth1
> Connect to socket: 0
> memory allocation on the socket: 0
> Link status: up
> Link speed: 10000 Mbps
> Link duplex: full-duplex
> MTU: 1500
> Promiscuous mode: enabled
> Allmulticast mode: disabled
> Maximum number of MAC addresses: 1
> Maximum number of MAC addresses of hash filtering: 0
> VLAN offload:
>   strip off
>   filter off
>   qinq(extend) off
> Hash key size in bytes: 40
> Supported RSS offload flow types:
>   ipv4
>   ipv4-frag
>   ipv4-tcp
>   ipv4-udp
>   ipv4-other
>   ipv6
>   ipv6-frag
>   ipv6-tcp
>   ipv6-udp
>   ipv6-other
>   user defined 15
>   user defined 16
>   user defined 17
> Minimum size of RX buffer: 0
> Maximum configurable length of RX packet: 1522
> Current number of RX queues: 1
> Max possible RX queues: 16
> Max possible number of RXDs per queue: 65535
> Min possible number of RXDs per queue: 0
> RXDs number alignment: 1
> Current number of TX queues: 1
> Max possible TX queues: 16
> Max possible number of TXDs per queue: 65535
> Min possible number of TXDs per queue: 0
> TXDs number alignment: 1
> 
> ********************* Infos for port 2  *********************
> MAC address: 00:0D:3A:38:09:DA
> Device name: net_failsafe_vsc1
> Driver name: net_failsafe
> Devargs: remote=eth2
> Connect to socket: 0
> memory allocation on the socket: 0
> Link status: up
> Link speed: 10000 Mbps
> Link duplex: full-duplex
> MTU: 1500
> Promiscuous mode: enabled
> Allmulticast mode: disabled
> Maximum number of MAC addresses: 1
> Maximum number of MAC addresses of hash filtering: 0
> VLAN offload:
>   strip off
>   filter off
>   qinq(extend) off
> Hash key size in bytes: 40
> Supported RSS offload flow types:
>   ipv4
>   ipv4-frag
>   ipv4-tcp
>   ipv4-udp
>   ipv4-other
>   ipv6
>   ipv6-frag
>   ipv6-tcp
>   ipv6-udp
>   ipv6-other
>   user defined 15
>   user defined 16
>   user defined 17
> Minimum size of RX buffer: 0
> Maximum configurable length of RX packet: 1522
> Current number of RX queues: 1
> Max possible RX queues: 16
> Max possible number of RXDs per queue: 65535
> Min possible number of RXDs per queue: 0
> RXDs number alignment: 1
> Current number of TX queues: 1
> Max possible TX queues: 16
> Max possible number of TXDs per queue: 65535
> Min possible number of TXDs per queue: 0
> TXDs number alignment: 1
> testpmd>  
> ```
> Two things of note about it.
> 
> 1. I have to specify `-w` option with a PCI ID of a card that **is not network card!** because otherwise testpmd binds my ssh interface and I lose command line. Above I specified PCI ID of a VGA adapter, so `testpmd` binds only ports specified on command line. If I don't spcify `-w`, `testmpd` binds a real PCI network card (since now DPDK is built with Mellanox driver included) and prevents any other packets from being transferred over my SSH connection.
> 2. For some reason second port got index `2` instead of `1`. Not sure why it happened.
> 
> I am inclined to close this issue because it became quite long already and I think most if not all of the problems were resolved. If you have any objections please commend or I will close it on the next week.
> 

1. The vdev_netvsc driver is supposed to take care of binding for you automatically.
It only binds devices that do not have any routes.  Whitelisting (as you found) is not going
to work as expected because of architecural differences:
  - the kernel network device (eth1) used by TAP is VMBUS based not PCI. Anyway TAP is not a
    bus based driver.
  - the VF device is hidden by failsafe, uses infiniband verbs, and is not directly useable anyway.

2. The index 1 is one of the two devices hidden behind failsafe.  Failsafe makes the two
   lower devices (which still exist in DPDK) hidden and only exposes the one combined device.

**Comment by gshimansky** [comment]
Closing this issue since it contains two solutions to bind ports in Azure: new one for netvsc and old one for failsafe. If you disagree, you can reopen it but I suggest opening a new bug with new details.
